dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/loadbalance/RoundRobinLoadBalance.java

```diff
@@ -32,72 +34,131 @@
+            methodWeightMap.putIfAbsent(key, new HashMap<String, WeightedRoundRobin>());
+            map = methodWeightMap.get(key);
         }
-        AtomicPositiveInteger sequence = sequences.get(key);
-        if (sequence == null) {
-            sequences.putIfAbsent(key, new AtomicPositiveInteger());
-            sequence = sequences.get(key);
-        }
-        int currentSequence = sequence.getAndIncrement();
-        if (maxWeight > 0 && minWeight < maxWeight) {
-            int mod = currentSequence % weightSum;
-            for (int i = 0; i < maxWeight; i++) {
-                for (Map.Entry<Invoker<T>, IntegerWrapper> each : invokerToWeightMap.entrySet()) {
-                    final Invoker<T> k = each.getKey();
-                    final IntegerWrapper v = each.getValue();
-                    if (mod == 0 && v.getValue() > 0) {
-                        return k;
-                    }
-                    if (v.getValue() > 0) {
-                        v.decrement();
-                        mod--;
+        synchronized (map) {
+            int totalWeight = 0;
+            int maxCurrent = Integer.MIN_VALUE;
+            long now = System.currentTimeMillis();
+            boolean needRecycle = false;
+            Invoker<T> selectedInvoker = null;
+            WeightedRoundRobin selectedWRR = null;
+            for (Invoker<T> invoker : invokers) {
+                String identifyString = invoker.getUrl().toIdentityString();
+                WeightedRoundRobin weightedRoundRobin = map.get(identifyString);
+                int weight = getWeight(invoker, invocation);
+                if (weight < 0) {
+                    weight = 0;
+                }
+                if (weightedRoundRobin == null) {
+                    weightedRoundRobin = new WeightedRoundRobin();
+                    weightedRoundRobin.setCurrent(0);
+                    weightedRoundRobin.setWeight(weight);
+                    weightedRoundRobin.setLastRecycle(now);
+                    map.put(identifyString, weightedRoundRobin);
+                }
+                if (weight != weightedRoundRobin.getWeight()) {
+                    //weight changed
+                    weightedRoundRobin.setCurrent(0);
+                    weightedRoundRobin.setWeight(weight);
+                }
+                int cur = weightedRoundRobin.increaseWeight();
+                weightedRoundRobin.setLastUpdate(now);
+                if (!needRecycle && now - weightedRoundRobin.getLastRecycle() > RECYCLE_PERIOD) {
+                    //try to recycle useless item every 60s if there is one outdated
+                    needRecycle = true;
+                }
+                if (cur > maxCurrent) {
+                    maxCurrent = cur;
+                    selectedInvoker = invoker;
+                    selectedWRR = weightedRoundRobin;
+                }
+                totalWeight += weight;
+            }
+            //recycle
+            if (needRecycle) {

```

Comment:
I suggestion use a out thread to check, not in this thread.

Reply:
Or ignore, cause not so much keys

Reply:
Amazing implementation. 

But we should avoid `synchronized` as much as possible, besides this synchronization block is too large to effect performance:

> Synchronized lock is an auto-upgrade lock from light weight lock in object header (when no concurrent) to monitor lock. So if the code block can be executed fast enough we may get a light weight lock mostly.

This is not true in this scenario, since it is very possible that multiple threads execute `doSelect` at the same moment.

I am fine with other parts, but this synchronization block is my major concern.